### PR TITLE
add boundary condition for edmfx

### LINKED ISSUE #1731: Add boundary condition for prognostic EDMFX

### PR DESCRIPTION
**Pull request by szy21**
## Purpose 

Closes #1731
co-authored-by: @trontrytel 

## Content

----
- [ ] I have read and checked the items on the review checklist.

### CODE REVIEW

File: src/cache/precomputed_quantities.jl [hunk 1]
```diff
@@ -366,7 +383,59 @@ function set_precomputed_quantities!(Y, p, t)
                     ᶜtsʲ,
                     ᶜspecificʲ.e_tot,
                 )
+
+                # Overwriting the first interior point for EDMFX ᶜh_totʲ
+                # based on boundary conditions for updrafts
+                ᶜh_tot_int = Fields.level(ᶜh_tot, 1)
+                ᶜh_totʲ_int = Fields.level(ᶜh_totʲ, 1)
+                @. ᶜh_totʲ_int = sgs_h_tot_first_interior_bc(
```
Comment (szy21) [inline src/cache/precomputed_quantities.jl:391]: I think you may need to do something like:
```
ᶜh_totʲ_int_value = Fields.field_values(ᶜh_totʲ_int)
@. ᶜh_totʲ_int_value = sgs_h_tot_first_interior_bc(...)
```
I'm surprised it doesn't work for `Field` though.

File: src/cache/precomputed_quantities.jl [hunk 2]
```diff
@@ -366,7 +383,59 @@ function set_precomputed_quantities!(Y, p, t)
...
+                    $Fields.field_values(ᶜz_int),
```
Comment (szy21) [inline src/cache/precomputed_quantities.jl:392]: Note to myself: double check if this should be `z_int` or `z_int - z_sfc`

Comment (trontrytel) [reply src/cache/precomputed_quantities.jl:392]: `z_int - z_sfc` looks more reasonable to me. Should I change it?

Comment (szy21) [reply src/cache/precomputed_quantities.jl]: Sounds good, thanks!

**Comment by trontrytel** [comment]
I think it is working now. I ended up turning everything into a `fields_value`, because I was still getting errors from the complainer complaining about spaces. It's not the most beautiful code, but maybe it gets the job done.

File: src/cache/precomputed_quantities.jl [hunk 3]
```diff
@@ -327,8 +327,13 @@ function set_precomputed_quantities!(Y, p, t)
     SurfaceConditions.update_surface_conditions!(Y, p, t)
 
     if turbconv_model isa EDMFX
+
+        #EDMFX BCs only support total energy as state variable
+        @assert energy_form isa TotalEnergy
```
Comment (szy21) [inline src/cache/precomputed_quantities.jl:332]: I think we are assuming `q_tot` exists too. Add assert moist model is not dry?

File: src/prognostic_equations/edmfx_boundary_condition.jl [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    sfc_buoyancy_flux,
+    sfc_ρ_flux_q_tot,
+    ustar,
+    obukhov_length,
+    sfc_local_geometry,
+) where {FT}
+    sfc_buoyancy_flux > 0 || return ᶜq_tot_int
+    # a_total = edmf.surface_area
+    # a_ = area_surface_bc(surface_conditions, edmf)
+    qt_var = get_first_interior_variance(
```
Comment (szy21) [inline src/prognostic_equations/edmfx_boundary_condition.jl:46]:
```suggestion
    q_tot_var = get_first_interior_variance(
```

File: src/prognostic_equations/edmfx_boundary_condition.jl [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        ustar,
+        ᶜz_int,
+        obukhov_length,
+        sfc_local_geometry,
+    )
+    # surface_scalar_coeff = percentile_bounds_mean_norm(
+    #     1 - a_total + (i - 1) * a_,
+    #     1 - a_total + i * a_,
+    # )
+    surface_scalar_coeff = FT(1.75)
```
[the remark this answers is not in the document]
Comment (trontrytel) [reply src/prognostic_equations/edmfx_boundary_condition.jl:29]: We want to do the easier thing first. And then maybe go back to the parameterised form

**Comment by trontrytel** [comment]
> The broadcasts look good to me, not sure if I should approve since @szy21 left some comments as well.

I'll apply them. But you could approve. @szy21 can't since she started the PR...

**Comment by szy21** [comment]
lol I approve that Nat can approve it

**Comment by trontrytel** [comment]
> lol I approve that Nat can approve it

I am grateful for your support ;)

**Pull request merged**